### PR TITLE
Throw exception when framework module tries to open an external URL 

### PR DESCRIPTION
**Pull request by Naktibalda**
#2328

Intercept click() for debugging
Throw exception when framework module tries to open an external URL

### CODE REVIEW

**Comment by Naktibalda** [comment]
Some Laravel4, Laravel5 and Yii2 tests failed, because they use domain based routes.
Some framework connectors don't support them, my guess is that this restriction should be limited to frameworks that remove http://localhost prefix from request URI:
https://github.com/Codeception/Codeception/search?utf8=%E2%9C%93&q=%24uri+%3D+str_replace%28%27http%3A%2F%2Flocalhost%27%2C+%27%27%2C+%24request-%3EgetUri%28%29%29%3B&type=Code

Phalcon, Universal, ZF1 and ZF2,

I think that a good solution would be to make them implement interface DoesntSupportDomains
or make others implement SupportsDomains

File: src/Codeception/Lib/InnerBrowser.php [hunk 1]
```diff
@@ -1338,4 +1346,20 @@ protected function redirectIfNecessary($result, $maxRedirects, $redirectCount)
...
     }
+
+    /**
+     * Clicks on a given link.
+     *
+     * @param Link $link A Link instance
+     *
+     * @return Crawler
+     */
+    public function clientClick(Link $link)
```
Comment (DavertMik) [inline src/Codeception/Lib/InnerBrowser.php]: it should not be public

**Comment by DavertMik** [comment]
@janhenkgerritsen @samdark what you think?
I think `SupportsDomains` can be implemented 

However, even domain-supported frameworks can't do redirects to external sites. But can we guess what sites are external and what are not?

**Comment by Naktibalda** [comment]
I think that such frameworks can handle it themselves.

**Comment by janhenkgerritsen** [comment]
I also think `SupportsDomains` can be implemented, but maybe change the name of the interface to `SupportsDomainRouting`?

For Laravel it will probably be possible to detect which URLs are external and which are not by simply looking through the registered routes and see if they have a domain specified. All domains that are specified can then be considered internal and all others external.

What currently happens for external routes with Laravel is that if there are no routes matching that domain the path of the URL wil be matched against the routes that are defined. This is actually very confusing behavior, because sometimes a match will be found and that page will be the response of the redirect (for example the homepage in case of a redirect to an URL without a path) and sometimes no match is found and you will get a 404 response. So this should be changed, but before I make any changes I will wait to see which solution will be implemented for this issue.

**Comment by Naktibalda** [comment]
any objections?

**Comment by janhenkgerritsen** [comment]
I still do think that `SupportsDomainRouting` or something like that would better communicate the intention of the interface, `SupportsDomains` is very generic.

And another suggestion, maybe give the interface a single method named something like `getInternalDomains` so the check for external redirects can be kept in a single location, instead of implementing it in each of the modules.

**Comment by DavertMik** [comment]
@samdark can we easily receive all internal domains in Yii? So to implement `getInternalDomains` method

**Comment by Naktibalda** [comment]
If you want to add a method to interface, it must be implemented in 6 framework modules.
And I can't do that.
It could be easier to implement it in each module.

**Comment by janhenkgerritsen** [comment]
Can't you just add a default implementation that returns an empty array? Then the maintainers can update the modules after you are finished. I can do the Laravel4,  Laravel5 and Lumen modules.

**Comment by Naktibalda** [comment]
@janhenkgerritsen empty array will break Laravel4 and Laravel5 tests:
https://travis-ci.org/Codeception/Codeception/jobs/80535452#L7134

 [ExternalUrlException] Codeception\Module\Laravel4 can't open external URL: http://l4.app/domain-route 

[ExternalUrlException] Codeception\Module\Laravel5 can't open external URL: http://myapp.com/posts/create

And we can't know what domains are used in tests of our users, so implementation with empty array can't be merged into 2.1 branch.

How about merging this pull request and then implementing getInternalDomains in a separate branch?

**Comment by janhenkgerritsen** [comment]
Merging this PR and implementing `getInternalDomains` in another branch is fine with me. Just notify me when you are finished with this PR and I will create the new branch and a first implementation of the changes.

**Comment by raistlin** [comment]
The more time I think about this, the more I dislike this.
Let me explain:
In my projects with Symfony2 I never set a domain in my Symfony2 code. I just setup apache virtual hosts to point the web path of my symfony2 app, and that is all. Testing domains are 'myapp.test', and production ones are 'myapp.com', application don't need to know on what domain is being running on.

In my testing environments, I just had configured in /etc/hosts files that 'myapp.test' points to 127.0.0.1, and in codeception.yml or acceptance.suite.yml I also have the same domain as base url for PhpBrowser and WebDriver. In production environment is setup simp

As far as I understand, you would like to intercept clicks of the tests to domains outside of what is defined in the framework, and raise an exception. 

What happens with people that have my same environment and never filter by hosts at their frameworks? Will I need to inject domains to my routes simply because of that feature? Did you first check if domain used is the same as stated in PhpBrowser or WebDriver? I just search for this fact but did not found this filtering. If you don't do this, then will force me to inject domain name to my framework, and I don't want to inject this just to prevent going outside my app while testing!

**Comment by janhenkgerritsen** [comment]
This functionality only applies in cases where you try to access an absolute URL through Codeception, for example by doing `$I->amOnPage('http://subdomain.example.com')`. But in an application that only runs on a single domain you won't do that, you will use only paths in your tests, for example `$I->amOnPage('contact')`.

You can take a look at the calling code here: https://github.com/Codeception/Codeception/blob/feature-get-internal-domains/src/Codeception/Lib/InnerBrowser.php#L1268

**Comment by raistlin** [comment]
Ok, then I'm ok with this.

Thanks for explanation, my mind was blowing :)

**Pull request merged**